buildSrc/src/main/groovy/io.deephaven.java-conventions.gradle

```diff
@@ -74,18 +76,85 @@ tasks.withType(JavaCompile).configureEach {
   }
 }
 
+def tier4CompilationExcludes = [
+  'io.deephaven.engine.table.impl.by.ssmpercentile.DoublePercentileTypeMedianHelper':['setResult'],
+  'io.deephaven.engine.table.impl.by.ssmpercentile.FloatPercentileTypeMedianHelper':['setResult'],
+  'io.deephaven.engine.table.impl.by.ssmpercentile.SsmChunkedPercentileOperator':['addChunk'],
+  'io.deephaven.engine.table.impl.by.IncrementalChunkedOperatorAggregationStateManager':['buildTable'],
+  'io.deephaven.engine.table.impl.RightIncrementalChunkedCrossJoinStateManager':['addToIndex', 'decorationProbe'],
+  'io.deephaven.engine.table.impl.IncrementalChunkedNaturalJoinStateManager':['buildTable', 'decorationProbe'],
+]
+def createCompilerDirectives = tasks.register('createCompilerDirectives') {
+  def c2ExcludesFile = project.layout.buildDirectory.file('c2-excludes.txt')
+  it.outputs.file(c2ExcludesFile)
+
+  doFirst {
+    def excludesAsList = []
+    tier4CompilationExcludes.forEach({type, methods ->
+      methods.forEach {method ->
+        excludesAsList += "${type}::${method}"
+      }
+    })
+    def builder = new JsonBuilder()
+    builder([{
+              match excludesAsList
+              c2 {
+                Exclude true
+              }
+            }])
+    c2ExcludesFile.get().asFile.text = builder.toPrettyString()
+  }
+}
+
+def c2ExcludesJvmArgs = { String c2ExcludesFile ->
+  return [
+          '-XX:+UnlockDiagnosticVMOptions',
+          "-XX:CompilerDirectivesFile=${c2ExcludesFile}",
+  ]
+}
+
+// Utility to add jvm args to all executions, whether intellij or from a application script or gradle javaexec
+def devJvmArgs = [
+//  '-XX:Tier4CompileThreshold=1000', // this optional line makes it easier to trigger the c2 error on the above methods
+//  '-XX:+PrintFlagsFinal',           // this optional line makes it easier to explore the final values for compiler args
+//  '-XX:+PrintCompilation',          // this optional line shows jit operations as they happen
+]
+
 tasks.withType(JavaExec).configureEach {
+  def c2ExcludesFile = createCompilerDirectives.get().outputs.files
+  inputs.files c2ExcludesFile
   javaLauncher.set runtimeLauncher
+  jvmArgs += c2ExcludesJvmArgs(c2ExcludesFile.singleFile.path) + devJvmArgs
 }
 
 tasks.withType(Test).configureEach {
+  def c2ExcludesFile = createCompilerDirectives.get().outputs.files
+  inputs.files c2ExcludesFile
+
   javaLauncher.set testRuntimeLauncher
+  jvmArgs += c2ExcludesJvmArgs(c2ExcludesFile.singleFile.path) + devJvmArgs
 }
 
 tasks.withType(GroovyCompile).configureEach {
   javaLauncher.set groovyCompilerLauncher
 }
 
+plugins.withType(ApplicationPlugin) {
+  applicationDistribution.into('lib') {
+    from(createCompilerDirectives.get().outputs.files)
+  }
+}
+tasks.withType(CreateStartScripts).configureEach {
+  inputs.properties(['scriptVers': Math.random()])
```

Comment:
Some form is, this is the dumb way of "never up to date". Ideally I show it that the classpath changes when the file changes, or we just put a static value here that we change when the file changes. I'll see if i can at least make buildSrc's own build.gradle emit some property or something that represents the contents of these files, since actually changing the file doesnt invalidate classpath (unlike changing _this_ file).

Reply:
Fixed, files can be read off the classpath as input, same as i'm providing them to the generator.